### DISCUSSION TITLE
Elsa Ideas

### DISCUSSION THREAD

**Discussion by sfmskywalker** [elsa-workflows/elsa-core#468] [Ideas]
This issue is a collection of pins for ideas to get feedback on, refine, prioritize & plan.

- [ ] Expression editor with intellisense
- [ ] Validation rules (e.g. mandatory properties) (see also #452)
- [ ] Test workflows from the designer using test inputs (including breakpoints)
- [ ] Define automated workflow tests from the dashboard & run them
- [ ] Workflow ACL rules: workflows can only be edited if one has the required permissions. Similar to Orchard Core content permissions.
- [ ] Restrict activities that are available from the designer. Perhaps configurable per workflow, or associated with a "whitelist profile" that itself has a set of required permissions a user must have (see also #315)
- [ ] See detailed diagnostic data during workflow development/testing (see also #444).
- [ ] Have telemetry data output during production workflow execution (e.g. to Application Insights).
- [ ] Modular Dashboard application. Could use (parts of) Blazing Orchard for dynamic component composition or even Oqtane. Allows developers to enrich the dashboard unobtrusively with application specific functionality.
- [ ] Pluggable workflow designer: configure per workflow what editor to use. Allows for optimized editing experience depending on the type of workflow you're editing. Alternatively, or perhaps ultimately, there would be a single designer that exposes knobs and bobs to configure it (also perhaps on a per workflow basis)
- [ ] Re-hostable workflow designer. Right now the designer is implemented using Blazor. Should be able to rehost it in e.g. Orchard Core, ABP Framework, Oqtane, etc. This allows develoeprs to embed the designer into their own backoffice applications.
- [ ] Conductor: a set of activities that allow for simple orchestration of tasks to be executed, similar to the way Netflix Conductor works. This allows Elsa to be used easily with any other type of application through a messaging transport system (HTTP, gRPC, AMQP) (#466)
- [ ] Workflow Templates: a way to design "master" workflows with "seams" into which "child" workflows can be inserted (#463)
- [ ] Workflows DSL / Code editor: the designer should have a toggle view that displays a code editor containing the DSL for the workflow being edited (#462)
- [x] Activity Providers: Currently, available activities are discovered from the DI container. By adding a layer of indirection, there could be more than one source of activities that yield activities. For instance, one provider would use DI to discover activity classes (as we have today), but other providers could for example parse en Open API specification and yield one activity per operation on a REST resource. So instead of having to manually configure SendHttpRequest activities to call out to some API such as GitHub, the toolbox would now have a category called GitHub and a list of activities to make API calls. Would work the same way for GraphQL endpoints, leveraging the schema to yield activities from. Another example is the Conductor style workflows, where users want to make calls to their own application. One could simply define a list of endpoints, and for each endpoint, an activity will automatically be made available. Yet another example might be applications using MassTransit: there could be an activity provider that yields activities based on the message consumers registered with DI. Each activity would represent sending/publishing a given message.
- [ ] Script Activities. There should be at least a JavaScript activity. A script activity allows you to implement arbitrary code that you would normally do using C#. This enables a lot of flexibility. And in combination with the Activity Providers feature, we could implement a dashboard module that allows the user to pre-define any number of scripted activities, which would become available as activities from the activity picker.
- [ ] Composite Activities: a way to create custom activities from the designer that are like mini workflows. This can currently be done using code already.
- [x] Re-introduce persistence abstractions. Right now in Elsa 2, everything depends on YesSQL directly. However, there is currently no need for this level of flexibility, and for applications that use Elsa in-process it is more convenient to be able to use the same data store for Elsa workflows and instances. For Elsa out-of-process (Elsa Server), this is less of an issue. (#496)
- [x] Service Bus activities. This is separate from the MassTransit activities. Rebus is currently used by Elsa intrinsically to schedule workflows to be executed (and picked up by consumers on any given node within a cluster). So I think it makes sense that if an application has workflows containing messaging / service bus activities, it would use the same message broker configured for Rebus. If not, we might introduce another messaging abstraction (similar to providing persistence abstractions) (#15)
- [ ] Add support for ABP Framework
- [ ] Multitenancy support. There should be easy to associate workflows with a tenant.
- [x] Activity types should be stored as part of ElsaOptions, rather than being added to the DI container, similar to what ORchard Core does with registering content part drivers and associated handlers. This might make for a more intuitive approach to adding activities to the system.
- [ ] Open API Activity Type Provider (dynamically expose activities representing HTTP requests supported by the Open API spec, e.g. GitHub's `ListUsers` and `FollowUser` REST API operations.
- [ ] GraphQL Activity Type Provider (dynamically expose activities representing GraphQL queries and mutations, e.g. GitHub's `GetUser`, `GetTopic` and `AddComment`.
- [ ] Perf: A smarter way to retrieve workflow instance summaries from the stores without first fetching the full documents.
- [ ] Add supporting APIs for the User Task activity + sample
- [ ] Raspberry PI activities
- [ ] CI/CD activities
- [ ] Kubernetes activities
- [ ] Saga / state machines
- [ ] Integration with [Temporal](https://docs.temporal.io/docs/workflows). Elsa for designing workflows, Temporal for driving the workflow

**Comment by xewn** [comment]
Hope it can be used directly in [AbpFramework](https://github.com/abpframework/abp)

**Comment by GeorgeS2019** [comment]
@xewn  thanks for highlighting AbpFramework

We may need to discuss in what ways the integration of Elsa will make best use of the following features of the AbpFramework

Examples

- [MULTI-TENANCY](https://docs.abp.io/en/abp/latest/Multi-Tenancy)
- [DOMAIN DRIVEN DESIGN INFRASTRUCTURE](https://docs.abp.io/en/abp/latest/Domain-Driven-Design)

Please share your views!!

**Reply by tunechr** [reply]
Yeah totally agree! 

Multi-tenancy def needed, hoping it can be done natively. and can work in a microservice approach as well. 

As you said I hope ABP incorporate more DDD and some CMS  style features(like content types in orchard/SharePoint) and command/eventing to their product.  So to be able to trigger workflows on entity type events and have entity-specific workflow actions... without having to code it all!   ABP is a really good product, so close to a great product!

**Comment by xyfy** [comment]
watched

**Reply by mrs01dev** [reply]
Ready Elsa Code Generator  For ABP :
https://github.com/EasyAbp/AbpHelper.CLI/
[https://github.com/EasyAbp/AbpHelper.CLI/](url)
The repository  AbpHelper.CLI include advanced  Elsa Code.
A dynamically generated workflow  is been created upon user input. 
Is there is a way to see the dynamically generated workflow in dashboard  ?
Is there is a way to invoke the dynamically generated workflow from   dashboard  ?

**Reply by sfmskywalker** [reply]
Yes, the Elsa 2 workflow dashboard will be able to display dynamically generated workflows, regardless of whether they are written using the Workflow Builder API or constructed dynamically using `WorkflowDefinition` / loaded from DB / parsed from JSON etc.

The dashboard will eventually also support invoking workflows, test them and debug them.
